Refactor Agent Portrayal to Use AgentPortrayalStyle Instead of Dictionary(#2436)

aarav-shukla07 opened the pull request:
### Summary
<!-- Provide a brief summary of the bug and its impact. -->
This PR refactors Mesa’s visualization system by replacing the dictionary-based agent_portrayal system with a new AgentPortrayalStyle class. This improves code clarity, ensures validation, and enhances backend compatibility across different visualization methods

### Bug / Issue
<!-- Link to the related issue(s) and describe the bug. Include details like the context, what was expected, and what actually happened. -->
Closes #2436

Previously, agent_portrayal functions returned a dictionary with properties like "color", "marker", "size", and "zorder". However, this approach:

- Lacked validation, allowing invalid properties.
- Made it harder to maintain and extend the portrayal system.
- Was inconsistent across Matplotlib and Altair visualizations.

This PR introduces AgentPortrayalStyle, making agent portrayal more structured and type-safe while maintaining backward compatibility.


### Implementation
<!-- Describe the changes made to resolve the issue. Highlight any important parts of the code that were modified. -->
- Created AgentPortrayalStyle class to encapsulate agent portrayal properties.
- Modified make_mpl_space_component to handle both AgentPortrayalStyle and dictionary-based portrayals for backward compatibility.
- Updated app.py examples to use AgentPortrayalStyle.
- Refactored test_call_space_drawer in test_solara_viz.py to ensure compatibility with the new portrayal system.
- Fixed space initialization issues in tests by correctly initializing ContinuousSpace.
- Added a seeded random generator to improve test reproducibility and remove warnings.


### Testing
Run all Mesa tests to ensure no regressions:
```
pytest
```
Verified that test_call_space_drawer correctly handles AgentPortrayalStyle
```
pytest tests/test_solara_viz.py::test_call_space_drawer -s
```
Result: All tests passed with no new failures.
### Additional Notes
- This PR maintains backward compatibility by allowing both dictionary-based portrayals and the new AgentPortrayalStyle class.
- Future improvements can involve updating all Mesa examples to exclusively use AgentPortrayalStyle.
- Fixing the test warnings about random number generation ensures consistent test results across runs.

EwoutH commented:
@aarav-shukla07 thanks for your PR! An `AgentPortrayalStyle` as a proper class would be great to have, I requested our visualization experts to review.

quaquel commented:
I like the basic idea and resulting API. However, I have some concerns about the implementation and the performance overhead that would require some custom benchmarking.

1. Can we make `AgentPortrayalStyle` a dataclass or something similar?
2. The new code in mpl_space_drawing iterates over the agents twice, which is wasteful. The new code is also largely not located in `collect_agent_data`, so I am struggling to understand what is going on here.
3. I am inclined to add the loc to `AgentPortrayalStyle`. This would nicely encapsulate all visual encoding in a single data object.

Sahil-Chhoker commented:
@aarav-shukla07, The code in `mpl_space_drawing.py` is same as in #2693 which is already closed. Please clean it up.

aarav-shukla07 commented:
> I like the basic idea and resulting API. However, I have some concerns about the implementation and the performance overhead that would require some custom benchmarking.
> 
> 1. Can we make `AgentPortrayalStyle` a dataclass or something similar?
> 2. The new code in mpl_space_drawing iterates over the agents twice, which is wasteful. The new code is also largely not located in `collect_agent_data`, so I am struggling to understand what is going on here.
> 3. I am inclined to add the loc to `AgentPortrayalStyle`. This would nicely encapsulate all visual encoding in a single data object.

Hi @quaquel ,

Thank you for your detailed feedback! Here’s how I’ve addressed your concerns:

(1) Converted AgentPortrayalStyle into a dataclass

This encapsulates all visual encoding in a single object, improving readability and structure.

(2) Optimized collect_agent_data

Previously, it iterated over agents twice. Now, it processes them in a single loop, reducing unnecessary overhead.

(3)  Added loc directly to AgentPortrayalStyle

This ensures that all relevant visual properties, including position, are encapsulated within a single structured object.

I’ve also verified that all tests pass , and I’ve resolved the merge conflicts to ensure smooth integration. Please let me know if there are any further improvements or clarifications needed!

Thanks again for your guidance. Looking forward to your feedback.

aarav-shukla07 commented:
Hi @Sahil-Chhoker ,

Thank you for your feedback. I acknowledge that the code in ``` mpl_space_drawing.py ``` was similar to the one in #2693, which has already been closed. I had cleaned up the redundant changes and ensure that only the necessary modifications remain in this PR.

I have updated the PR  and verified that the changes align with the latest codebase. Please let me know if there are any further improvements or clarification needed!

Thanks again for your guidance. Looking forward to your feedback.

quaquel commented on mesa/visualization/AgentPortrayalStyle.py:
```diff
@@ -0,0 +1,28 @@
+    edgecolors: str = "black"  # edge color for markers
+    loc: tuple[float, float] | None = None  # stores agent position
+
+    def to_dict(self):
```
I would suggest doing it the other way arround. In get_agent_data if the return is a dict cast it to a AgentPortrayalStyle and issue a warning

quaquel commented on mesa/visualization/components/matplotlib_components.py:
```diff
@@ -46,12 +47,20 @@ def make_mpl_space_component(
+            return AgentPortrayalStyle()
 
     def MakeSpaceMatplotlib(model):
+        def wrapped_agent_portrayal(agent):
```
can't this be handled inside get_agent_data?

quaquel commented on mesa/visualization/mpl_space_drawing.py:
```diff
@@ -47,69 +47,31 @@
+        portrayal_dict = agent_portrayal(agent) or {}
+
+        agent_data_list.append(
+            AgentPortrayalStyle(
```
This is wasteful because in the current code you already allways get an `AgentPortrayalStyle` instance.

quaquel commented on mesa/visualization/mpl_space_drawing.py:
```diff
@@ -47,69 +47,31 @@
+                alpha=portrayal_dict.get("alpha", 1.0),
+                linewidths=portrayal_dict.get("linewidths", 1.0),
+                edgecolors=portrayal_dict.get("edgecolors", "black"),
+                loc=agent.pos if agent.pos is not None else (0, 0),  # Store location
```
this is not correct.

loc is either agent.pos or agent.cell.coordinate

quaquel commented on mesa/visualization/mpl_space_drawing.py:
```diff
@@ -346,7 +308,9 @@ def draw_orthogonal_grid(
     s_default = (180 / max(space.width, space.height)) ** 2
-    arguments = collect_agent_data(space, agent_portrayal, size=s_default)
+    arguments = collect_agent_data(
+        space, agent_portrayal, default_color="tab:red", default_size=30
```
these defaults can be moved into `AgentPortrayalStyle`

quaquel commented on mesa/visualization/mpl_space_drawing.py:
```diff
@@ -47,69 +47,31 @@
-    arr[:] = arguments["marker"]
-    data["marker"] = arr
-    return data
+    return agent_data_list
```
not sure but it might be better to turn all agent data into numpy arrays already here so we avoid needless iterations later.

quaquel commented on mesa/visualization/mpl_space_drawing.py:
```diff
@@ -483,25 +434,39 @@ def draw_network(
+            agent_style.loc = (0, 0)  # Fallback to (0,0) if node not found
+
+    # Convert data for plotting
+    scatter_data = {
+        "s": [a.size for a in agent_data_list],
+        "c": [a.color for a in agent_data_list],
+        "marker": [a.marker for a in agent_data_list],
+        "zorder": [a.zorder for a in agent_data_list],
+        "loc": np.array([a.loc for a in agent_data_list], dtype=np.float64),
+        "alpha": [a.alpha for a in agent_data_list],
+        "edgecolors": [a.edgecolors for a in agent_data_list],
+        "linewidths": [a.linewidths for a in agent_data_list],
+    }
```
this is wasteful iteration. You iterate over the same data 8 times.

quaquel commented on mesa/visualization/mpl_space_drawing.py:
```diff
@@ -623,40 +592,46 @@ def _scatter(ax: Axes, arguments, **kwargs):
-                    f"{entry} is specified in agent portrayal and via plotting kwargs, you can only use one or the other"
-                )
+
+    marker = [agent_data.marker for agent_data in arguments]
+    zorder = [agent_data.zorder for agent_data in arguments]
+    edgecolors = [agent_data.edgecolors for agent_data in arguments]
+    linewidths = [agent_data.linewidths for agent_data in arguments]
+    alpha = [agent_data.alpha for agent_data in arguments]
```
same here, 5 repeated iterations

quaquel reviewed (changes requested):
There are various things that can be done to improve this PR further. Let me know if you have questions.

aarav-shukla07 commented on mesa/visualization/mpl_space_drawing.py:
```diff
@@ -386,19 +350,23 @@ def draw_hex_grid(
         loc[:, 0] = loc[:, 0] * x_spacing + ((loc[:, 1] - 1) % 2) * (x_spacing / 2)
         loc[:, 1] = loc[:, 1] * y_spacing
-        arguments["loc"] = loc
+        for i, agent_data in enumerate(arguments):
```
Previously, ```arguments``` was structured as a dictionary holding arrays for all agents — so we could assign the entire ```loc``` array at once using ```arguments["loc"] = loc.```

With the updated structure, ```arguments``` is now a list of dictionaries, each representing an individual agent’s data. To assign each agent its own specific location, we iterate through ```arguments``` and update each ```agent_data entry:
```bash
for i, agent_data in enumerate(arguments):
    agent_data["loc"] = loc[i]
```
This change ensures that each agent dictionary gets its corresponding location from the ```loc``` array, maintaining alignment between spatial data and individual agent portrayal.

EwoutH commented:
@Sahil-Chhoker or @sanika-n, could (one of) you review this PR?

aarav-shukla07 commented:
Hi @quaquel ,

Thank you for your valuable feedback! 
Based on your suggestions, I’ve made the following changes:

- Consolidated multiple iterations over ```agent_data_list``` into a single pass while constructing ```scatter_data```.

- Refactored ```collect_agent_data``` to return NumPy arrays directly, avoiding unnecessary intermediate structures.

- Ensured compatibility across all grids and resolved the attribute issues that appeared during testing.

All tests are now passing locally after resolving the merge conflict. Please let me know if you’d prefer further optimizations (e.g., using structured arrays or different handling of defaults).

Looking forward to your thoughts!

Sahil-Chhoker commented:
> @Sahil-Chhoker / @sanika-n, could (one of) you review this PR?

Really sorry again, I am travelling most of the next week (until 10th) so I won't be able to find time to review this PR, @sanika-n is it alright if I leave the review to you?

sanika-n commented:
I am currently tied up with my proposal, but I can take this up after 8th

Sahil-Chhoker reviewed (changes requested):
Hey @aarav-shukla07, I took a little look at this PR, so to just start the review can you please start by incorporating quaquel's review? Also I see that the previous docstrings are all missing, I don't see any need to remove them, so fix that as well. And at last, can you resolve the merge conflicts?

sanika-n commented:
Also, could you pls update the tests/test_components_matplotlib.py file as well? That still uses a dictionary for agent-portrayal, you can find the code [here](https://github.com/projectmesa/mesa/blob/b490a6638b6f5d1a7ce0bc708e18b9ec35aae72f/tests/test_components_matplotlib.py#L31C1-L44C1)

aarav-shukla07 commented:
Hello @Sahil-Chhoker  @sanika-n 
Since I was travelling for the last few days, so I wasn't able to changes the code, so sorry for that.
I will try my best to update the changes as fast as possible.
Thank you

EwoutH commented:
Hi @aarav-shukla07! Are you interested in continuing working on this PR?
